march_hardware_interface/src/march_hardware_interface.cpp

```diff
@@ -247,6 +249,36 @@ void MarchHardwareInterface::write(ros::Duration elapsed_time)
   }
 }
 
+void MarchHardwareInterface::resetIMotionCubesUntilTheyWork()
+{
+  bool encoderSetCorrectly = false;
+
+  while (!encoderSetCorrectly)
+  {
+    encoderSetCorrectly = true;
+    for (int i = 0; i < num_joints_; ++i)
+    {
+      march4cpp::Joint joint = marchRobot.getJoint(joint_names_[i]);
+      if (joint.getAngleIU() == 0)
+      {
+        ROS_ERROR("Joint %s failed (encoder reset)", joint_names_[i].c_str());
+        encoderSetCorrectly = false;
+      }
+    }
+    for (int i = 0; i < num_joints_; ++i)
```

Comment:
Why do we reset all the IMC's while only one of them displays an angle of 0? Could we implement the IMC reset in the if loop where we check whether the angle is 0?

Reply:
Hmmm... It is probably because you want to restart the ethercat when one of the IMC's has been reset. And you only want to do it when the all the IMC's are reset and not after every one of them.

Reply:
I think that this piece of code could be more efficient, but I do not directly see how. If you are okay with it, I think you could merge it.